File: N-Queens/nQueens.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<string> > solveNQueens(int n) {
        vector<vector<string> > res;
        vector<string> nQueens(n, string(n, '.'));
        solveNQueensHelper(res, nQueens, 0, n);
        return res;
    }
    void solveNQueensHelper(vector<vector<string> >& res, vector<string>& nQueens, int row, int& n) {
        if (row == n) {
            res.push_back(nQueens);
            return;
        }
        for (int col = 0; col < n; ++col) {
            if (isValid(nQueens, row, col, n)) {
                nQueens[row][col] = 'Q';
                solveNQueensHelper(res, nQueens, row + 1, n);
                nQueens[row][col] = '.';
            }
        }
    }
    bool isValid(vector<string>& nQueens, int row, int col, int& n) {
        for (int i = 0; i < row; i++) {
            if (nQueens[i][col] == 'Q')
                return false;
        }
        // 45 degree diagonal in preLevels
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; --i, --j) {
            if (nQueens[i][j] == 'Q')
                return false;
        }
        // 135 degree diagonal in preLevels
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; --i, ++j) {
            if (nQueens[i][j] == 'Q')
                return false;
        }
        return true;
    }
};

int main()
{
  Solution s;
  for (vector<string>& v : s.solveNQueens(8)) {
    for (string i : v)
      cout << i << " ";
    cout << endl;
  }
  return 0;
}

File: N-Queens/nQueens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }
        dfs(n, board, res, 0);
        return res;
    }
    
    public void dfs(int n, char[][] board, List<List<String>> res, int row) {
        if (row == n) {
            res.add(construct(board));
            return;
        }
        for (int col = 0; col < n; col++) {
            if (isValid(board, row, col, n)) {
                board[row][col] = 'Q';
                dfs(n, board, res, row + 1);
                board[row][col] = '.';
            }
        }
    }
    
    private boolean isValid(char[][] board, int row, int col, int n) {
        //check if the column had a queen before.
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 'Q') return false;
        }
        //check if the 45° diagonal had a queen before.
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') return false;
        }
        //check if the 135° diagonal had a queen before.
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') return false;
        }
        return true;
    }
    
    private List<String> construct(char[][] board) {
        List<String> res = new ArrayList<String>();
        for(int i = 0; i < board.length; i++) {
            String s = new String(board[i]);
            res.add(s);
        }
        return res;
    }
}
